File: src/Fantomas.VisualStudio/Commands/FormatDocumentCommand.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Hestia.FSharpCommands.Utils;
using Microsoft.VisualStudio.Shell;
using Microsoft.VisualStudio.Text;
using Microsoft.VisualStudio.Text.Editor;
using Microsoft.VisualStudio.Text.Formatting;

namespace Hestia.FSharpCommands.Commands
{
    public class FormatDocumentCommand : FormatCommand
    {
        public override void Execute()
        {
            using (Cursor.Wait())
            {
                ExecuteFormat();
            }
        }

        protected override string GetFormatted(bool isSignatureFile, string source, Fantomas.FormatConfig.FormatConfig config)
        {        
            return Fantomas.CodeFormatter.formatSourceString(isSignatureFile, source, config);
        }

        protected override Action GetNewCaretPositionSetter()
        {
            var currentSnapshot = TextView.TextBuffer.CurrentSnapshot;

            var caretPos = TextView.Caret.Position.BufferPosition;
            var caretLine = currentSnapshot.GetLineFromPosition(caretPos.Position);
            int line = currentSnapshot.GetLineNumberFromPosition(caretPos);
            int column = caretPos - caretLine.Start;
            
            // Get start line of scroll bar
            var scrollBarLine = TextView.TextViewLines.FirstOrDefault(l => l.VisibilityState != VisibilityState.Hidden);
            int scrollBarPos = (scrollBarLine == null) ? 0 : currentSnapshot.GetLineNumberFromPosition(scrollBarLine.Start);
            int maxLine = currentSnapshot.LineCount;

            Action setNewCaretPosition = () =>
                {
                    var newCurrentSnapshot = TextView.TextBuffer.CurrentSnapshot;
                    int newMaxLine = newCurrentSnapshot.LineCount;

                    // Scale caret positions in a linear way
                    int newLine = (int) ((float)line * (float)newMaxLine / (float)maxLine);
                    var newCaretPos = newCurrentSnapshot.GetLineFromLineNumber(newLine).Start.Add(column);
                    var caretPoint = new VirtualSnapshotPoint(newCurrentSnapshot, newCaretPos);
                    TextView.Caret.MoveTo(caretPoint);

                    // Assume that the document scales in a linear way
                    int newScrollBarPos = (int) ((float)scrollBarPos * (float)newMaxLine / (float)maxLine);
                    TextView.ViewScroller.ScrollViewportVerticallyByLines(ScrollDirection.Down, newScrollBarPos);
                };

            return setNewCaretPosition;
        }
    }
}

File: src/Fantomas.UI/FormatDocumentCommand.fs
namespace FSharpVSPowerTools.CodeFormatting

open System
open System.Collections.Generic
open System.IO
open System.Linq
open System.Text
open System.Threading.Tasks
open Microsoft.VisualStudio.Shell
open Microsoft.VisualStudio.Text
open Microsoft.VisualStudio.Text.Editor
open Microsoft.VisualStudio.Text.Formatting
open Fantomas.FormatConfig
open Fantomas.CodeFormatter

type FormatDocumentCommand(getConfig: Func<FormatConfig>) =
    inherit FormatCommand(getConfig)

    override x.Execute() =
        use disposable = Cursor.wait()
        x.ExecuteFormat()

    override x.GetFormatted(isSignatureFile: bool, source: string, config: FormatConfig) =
        formatSourceString isSignatureFile source config

    override x.SetNewCaretPosition(caretPos, scrollBarPos, originalSnapshot) =
        let caretLine = originalSnapshot.GetLineFromPosition(caretPos.Position)
        let line = originalSnapshot.GetLineNumberFromPosition(int caretPos)
        let column = caretPos - caretLine.Start
        let maxLine = originalSnapshot.LineCount

        let currentSnapshot = x.TextBuffer.CurrentSnapshot
        let newMaxLine = currentSnapshot.LineCount

        // Scale caret positions in a linear way
        let newLineNo = int (float line * (float newMaxLine) / (float maxLine))
        let newLine = currentSnapshot.GetLineFromLineNumber(newLineNo)
        let newColumn = min column newLine.Length
        let newCaretPos = newLine.Start.Add(newColumn)
        let caretPolet = VirtualSnapshotPoint(currentSnapshot, int newCaretPos)
        x.TextView.Caret.MoveTo(caretPolet) |> ignore

        // Assume that the document scales in a linear way
        let newScrollBarPos = int (float scrollBarPos * (float newMaxLine) / (float maxLine))
        x.TextView.ViewScroller.ScrollViewportVerticallyByLines(ScrollDirection.Down, newScrollBarPos)